test/experimental/eof_validation.cpp

```diff
@@ -0,0 +1,420 @@
+
+    initcode_failure,
+    impossible,
+};
+
+struct ExecutionMock
+{
+    std::variant<bytes, error_code> execution_result;
+
+    std::variant<bytes, error_code> execute([[maybe_unused]] bytes_view code)
+    {
+        return execution_result;
+    }
+};
+
+int determine_eof_version(bytes_view code);
+error_code validate(bytes_view code, int expected_version);
+error_code validate_eof0(bytes_view code);
+error_code validate_eof1(bytes_view code_without_prefix);
+std::variant<bytes, error_code> create_contract_v1(
+    ExecutionMock& ee, bytes_view initcode, int eof_version);
+std::variant<bytes, error_code> create_contract_v2(
+    ExecutionMock& ee, bytes_view initcode, int eof_version);
+std::variant<bytes, error_code> execute_create_tx_v1(ExecutionMock& ee, bytes_view initcode);
+std::variant<bytes, error_code> execute_create_tx_v2(ExecutionMock& ee, bytes_view initcode);
+
+/// Determine the EOF version of the code by inspecting code's EOF prefix.
+/// If the prefix is missing or invalid, the 0 is returned meaning legacy code.
+int determine_eof_version(bytes_view code)
+{
+    return (code.size() >= 4 && code[0] == FORMAT && code[1] == MAGIC[0] && code[2] == MAGIC[1]) ?
+               code[3] :
+               0;
+}
+
+error_code validate(bytes_view code, int expected_version)
+{
+    const auto version = determine_eof_version(code);
+    if (version != expected_version)
+        return error_code::eof_version_mismatch;
+
+    auto code_without_prefix = code;
+    code_without_prefix.remove_prefix(4);
+    switch (version)
+    {
+    default:
+        return error_code::eof_version_unknown;
+    case 0:
+        return validate_eof0(code);
+    case 1:
+        return validate_eof1(code_without_prefix);
+    }
+}
+
+error_code validate_eof0(bytes_view code)
+{
+    return (code.empty() || code[0] != FORMAT) ? error_code::success :
+                                                 error_code::starts_with_format;
+}
+
+error_code validate_eof1(bytes_view code_without_prefix)
+{
+    enum class State
+    {
+        section_id,
+        section_size,
+        terminated
+    };
+
+    auto state = State::section_id;
+    uint8_t section_id;
+    int section_sizes[3] = {0, 0, 0};
+    auto it = code_without_prefix.begin();
+    while (it != code_without_prefix.end() && state != State::terminated)
+    {
+        switch (state)
+        {
+        case State::section_id:
+        {
+            section_id = *it;
+            switch (section_id)
+            {
+            case TERMINATOR:
+                if (section_sizes[CODE_SECTION] == 0)
+                    return error_code::code_section_missing;
+                state = State::terminated;
+                break;
+            case DATA_SECTION:
+                if (section_sizes[CODE_SECTION] == 0)
+                    return error_code::code_section_missing;
+                [[fallthrough]];
+            case CODE_SECTION:
+                state = State::section_size;
```

Comment:
Invalid case of multiple code/data sections is not handled here. Fixed in https://github.com/ethereum/evmone/pull/334/commits/0995f07141bc82204f4f5fd97c5fc84da0ca75f4